src/plugins/lua/replies.lua

```diff
@@ -34,42 +34,102 @@ local settings = {
   expire = 86400, -- 1 day by default
   key_prefix = 'rr',
   key_size = 20,
+  sender_prefix = 'rsrk',
+  sender_key_global = 'verified_senders',
+  sender_key_size = 20,
   message = 'Message is reply to one we originated',
   symbol = 'REPLY',
   score = -4, -- Default score
+  max_local_size = 20,
+  max_global_size = 30,
   use_auth = true,
   use_local = true,
   cookie = nil,
   cookie_key = nil,
   cookie_is_pattern = false,
   cookie_valid_time = '2w', -- 2 weeks by default
   min_message_id = 2, -- minimum length of the message-id header
+  reply_sender_privacy = false,
+  reply_sender_privacy_alg = 'blake2',
+  reply_sender_privacy_prefix = 'obf',
+  reply_sender_privacy_length = 16,
 }
 
 local N = "replies"
 
 local function make_key(goop, sz, prefix)
   local h = hash.create()
   h:update(goop)
-  local key
-  if sz then
-    key = h:base32():sub(1, sz)
-  else
-    key = h:base32()
-  end
+  local key = (prefix or '') .. h:base32():sub(1, sz)
+  return key
+end
 
-  if prefix then
-    key = prefix .. key
-  end
+local global_replies_set_script
+local local_replies_set_script
+
+local function configure_redis_scripts(_, _)
+  local redis_script_zadd_global = [[
+      -- getting last score of the recipients from the global replies set
+      local data = {}
+      data = redis.call('ZRANGE', KEYS[1], '-1', '-1', 'WITHSCORES')
+      local score = tonumber(data[#data])
+      if score == nil then
+        score = 0
+      end
 
-  return key
+      local global_size = redis.call('ZCARD', KEYS[1])
+      if global_size > {= mgs =} then
+        redis.call('ZREMRANGEBYRANK', KEYS[1], '{= mgs =}', global_size)
+      end
+
+      local recipients_addrs = {}
+      recipients_addrs = ARGV
+      if recipients_addrs ~= nil then
+        for _, rcpt in ipairs(recipients_addrs) do
+          -- score is using here as a one-based numeration just like in lua
+          score = score + 1
+          -- adding recipients to the global replies set
+          redis.call('ZADD', KEYS[1], score, rcpt)
+        end
+      end
+      ]]
+  local set_script_zadd_global = lua_util.jinja_template(redis_script_zadd_global,
+          { mgs = settings['max_global_size'] })
+  rspamd_logger.debugm(N, rspamd_config, 'added add to global replies set script %s', set_script_zadd_global)
+  global_replies_set_script =  lua_redis.add_redis_script(set_script_zadd_global, redis_params)
+
+  local redis_script_zadd_local = [[
+      local local_replies_set_size = redis.call('ZCARD', KEYS[1])
+      if local_replies_set_size > {= mls =} then
+        redis.call('ZREMRANGEBYRANK', KEYS[1], '{= mls =}', local_replies_set_size)
+      end
+      local params = {}
+      params = ARGV
+      if params ~= nil then
+          local task_time = params[1]
+          table.remove(params, 1)
+
+          for _, rcpt in ipairs(params) do
+            -- adding recipients for the local replies set
+            redis.call('ZADD', KEYS[1], task_time, rcpt)
```

Comment:
We could add more arguments to `ZADD` rather than using `for` loop (but it's a little unpleasant to do).